### DISCUSSION TITLE
Insufficient type overlap for parametrized types like `UuidFilter<$PrismaModel>`

### DISCUSSION THREAD

**Discussion by pmlk** [prisma/prisma#22858] [Q&A]
### Bug description

When I define two models, e.g. `UserA` and `UserB` with the exact same field names and types, e.g. `id String @db.Uuid`, the resulting `User<A|B>WhereInput` types do not overlap sufficiently. This causes my IDE to complain with 

`Type 'UserAWhereInput | UserBWhereInput' is not assignable to type 'UserAWhereInput'.` or   
`Type 'UserAWhereInput | UserBWhereInput' is not assignable to type 'UserBWhereInput'.` 

(see code example below).

I dug my way through the types. From `User<A|B>WhereInput` via `id?: UuidFilter<"UserA"> | string;` (or `id?: UuidFilter<"UserB"> | string;`) I eventually end up at this definition in `@prisma/client/runtime/library.d.ts`:

```ts
export declare interface FieldRef<Model, FieldType> {
    readonly modelName: Model;
    readonly name: string;
    readonly typeName: FieldType;
    readonly isList: boolean;
}
```

The culprit seems to be the `Model` part, which will evaluate to the string `"UserA"` or `"UserB"`, thus preventing `UserAWhereInput` and `UserBWhereInput` from overlapping.

Note that the same applies to `DateTimeFilter` and other parametrizable types where the model name is passed as a type parameter.

Before upgrading to `v5` in `v4.11.0` these type were not parametrizable, so it was just `UuidFilter` instead of `UuidFilter<$PrismaModel>`, and that didn't cause this issue.

Workaround for now is to pass `where` with `as`: `where as Prisma.UserAWhereInput`, but that comes with other potential issues once `UserA` and `UserB` diverge and the tooling won't alert us of us doing something wrong.

### How to reproduce

See prisma schema and prisma client queries below.

### Expected behavior

In the example, the resulting `UserAWhereInput` and `UserBWhereInput` types should sufficiently overlap so that `where: UserAWhereInput | UserBWhereInput` can be passed where only `UserAWhereInput` (or only `UserBWhereInput`) is expected without the typing system complaining with

`Type 'UserAWhereInput | UserBWhereInput' is not assignable to type 'UserAWhereInput'.` or
`Type 'UserAWhereInput | UserBWhereInput' is not assignable to type 'UserBWhereInput'.`

### Prisma information

```prisma
model UserA {
  id String @id() @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
}

model UserB {
  id String @id() @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
}
```

```ts
async getUser(where: Prisma.UserAWhereInput | Prisma.UserBWhereInput, aOrB: 'a' | 'b'): Promise<UserA | UserB> {
  if (aOrB === 'a') {
    return this.prismaService.userA.findFirstOrThrow({ where });
  } else {
    return this.prismaService.userB.findFirstOrThrow({ where });
  }
}
```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v20.7.0

### Prisma Version

```
prisma                  : 5.8.1
@prisma/client          : 5.8.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.7.0
Query Engine (Node-API) : libquery-engine 78caf6feeaed953168c64e15a249c3e9a033ebe2 (at node_modules/prisma/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 78caf6feeaed953168c64e15a249c3e9a033ebe2 (at node_modules/prisma/node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.8.1-1.78caf6feeaed953168c64e15a249c3e9a033ebe2
Default Engines Hash    : 78caf6feeaed953168c64e15a249c3e9a033ebe2
Studio                  : 0.497.0
Preview Features        : multiSchema, postgresqlExtensions, relationJoins
```

**Comment by SevInf** [comment]
As a part of the filter types, models accept [references to other fields in the same model](https://www.prisma.io/docs/orm/reference/prisma-client-reference#compare-columns-in-the-same-table). "Same model" is the key here: `prisma.userA.findFirst({ where: { id: prisma.userB.fields.id } })` won't be valid, even if `UserA` and `UserB` have identical fields. Therefore, `UserAInput` is not compatible with `UserBInput`.

**Comment by pmlk** [comment]
That seems to be a useful feature. Without having had a deeper look, I'm not sure if this requires the `<Model>WhereInput` types to be defined this way, though. But regardless, I can only provide my feedback that it breaks our workflow and it makes Union Types with `<Model>WhereInput` types, which we found really useful, essentially evaluate to `never`.

I really like Prisma because of its strong typing, but I think this goes one step too far, where types are now strongly coupled to the model it comes from. It makes Union Types unusable and interface segregation with the auto generated types impossible. The workaround with `as` circumvents the type system and thus one of my favorite features of Prisma: its type safety.

**Comment by SevInf** [comment]
>That seems to be a useful feature. Without having had a deeper look, I'm not sure if this requires the <Model>WhereInput types

It does (or requires something similar) because arbitrary `UserAWhereInput`   is not assignable `UserBWhereInput` even if the two have identical fields. `{ id: prisma.userA.fields.id }` is an example of valid `UserAWhereInput`  that is not assignable to  `UserBWhereInput`. Incompatibility between those 2 types is not a bug, it is something we want.

**Comment by SevInf** [comment]
What you can do in this case, is instead of intersection between `UserAWhereInput` and `UserBWhereInput` (which semantically means value is both at the same time, which we established before is not possible) use something like "common fields between userA and userB", which would automatically exclude field references and all other corner cases where types would be different.  [This answer](https://stackoverflow.com/a/47375979) could be a good starting point for implementing such type.

**Comment by aqrln** [comment]
Since this is working as intended, I'll go ahead and convert this issue to a discussion to keep the backlog clean.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
